core/arch/arm/mm/sp_mem.c

```diff
@@ -0,0 +1,196 @@
+ */
+#include <assert.h>
+#include <ffa.h>
+#include <kernel/spinlock.h>
+#include <mm/fobj.h>
+#include <mm/mobj.h>
+#include <mm/sp_mem.h>
+
+/*
+ * nw_pa is used to map the offset used with the mobj to the Normal World
+ * physical address.
+ */
+struct nw_pa {
+	paddr_t pa;
+	size_t offset;
+
+	SLIST_ENTRY(nw_pa) link;
+};
+
+/*
+ * nw_pa_list stores all physical address of the share coming from Normal World.
+ * It is used to get the physical address of a Normal World share
+ */
+SLIST_HEAD(nw_pa_head_t, nw_pa);
+static struct nw_pa_head_t nw_pa_list = SLIST_HEAD_INITIALIZER(nw_pa_head_t);
+/*
+ * We use 1 mobj and 1 fobj to for mapping Normal World memory that is shared
+ * with one of the SP endpoints.
+ */
+static struct fobj nw_fobj;
+static struct mobj *nw_mobj;
+
+/* mem_shares stores all active FF-A shares. */
+SLIST_HEAD(sp_mem_head, sp_mem);
+static struct sp_mem_head mem_shares = SLIST_HEAD_INITIALIZER(sp_mem_head);
+
+static void sp_mem_free_fobj(struct fobj *fobj __unused);
+static paddr_t get_pa(struct fobj *obj, uint32_t page_idx);
+
+static unsigned int sp_mem_lock = SPINLOCK_UNLOCK;
+
+const struct fobj_ops ops_sp_mem  __weak __rodata_unpaged("ops_sp") = {
+	.free = sp_mem_free_fobj,
+	.get_pa = get_pa,
+};
+
+static void sp_mem_free_fobj(struct fobj *fobj __unused)
+{}
+
+static paddr_t get_pa(struct fobj *obj __maybe_unused, uint32_t page_idx)
+{
+	struct nw_pa *pa_offset = NULL;
+
+	SLIST_FOREACH(pa_offset, &nw_pa_list, link) {
+		if (pa_offset->offset == page_idx)
+			return pa_offset->pa;
+	}
+	return 0;
+}
+
+struct sp_mem *sp_mem_new(void)
+{
+	struct sp_mem *smem = NULL;
+	uint32_t exceptions = 0;
+	static uint64_t global_handle = 1;
+
+	smem = malloc(sizeof(*smem));
+	if (!smem)
+		return NULL;
+
+	exceptions = cpu_spin_lock_xsave(&sp_mem_lock);
+	/*
+	 * OP_TEE SHAREs use bit 44 use bit 45 instead.
+	 */
+	smem->transaction.global_handle = global_handle | BIT64(45);
+	global_handle++;
+	cpu_spin_unlock_xrestore(&sp_mem_lock, exceptions);
+
+	SLIST_INIT(&smem->regions);
+	SLIST_INSERT_HEAD(&mem_shares, smem, link);
+
+	return smem;
+}
+
+static size_t nw_new_offset(void)
+{
+	struct nw_pa *last = NULL;
+	struct nw_pa *pa_offset = NULL;
+	bool found = true;
+	size_t offset = 0;
+	uint32_t i = 0;
+
+	last = SLIST_FIRST(&nw_pa_list);
+	if (!last)
+		return offset;
+
+	offset = last->offset;
+
+	/* Make sure we have a unique offset value*/
+	while (found) {
+		/*
+		 * The offset has to be aligned with the page size. get_pa() is
+		 * called with the offset masked with the page size.
+		 */
+		offset += SMALL_PAGE_SIZE;
+		found = false;
+		SLIST_FOREACH(pa_offset, &nw_pa_list, link) {
+			if (pa_offset->offset == offset) {
+				found = true;
+				break;
+			}
+		}
+		/*Make sure we don't loop forever*/
+		i++;
+		assert(i < (UINT32_MAX / SMALL_PAGE_SIZE));
+	}
+
+	return offset;
+}
+
+struct mobj *sp_mem_get_nw_mobj(paddr_t pa, size_t *offset)
+{
+	struct nw_pa *nw_pa = malloc(sizeof(struct nw_pa));
+
+	if (!nw_mobj) {
```

Comment:
In the end we are using vm_map to map the region. vm_map does require a mobj. Or do you think it would be a better approach to create a separate function instead of vm_map which doesn't use any mobj objects?
This doesn't seem to straight forward to me as umap_add_region  uses the `reg->mobj->size ` unless we just add add the size to it

Reply:
I meant to add a new mobj/fobj to describe the memory of this FFA_MEM_SHARE, where `mobj->size` is the actual size of the shared memory. That seems a much more straightforward solution with all the information in one place (ok, maybe two in case you use a fobj too) instead of scattered in many structs. `mobj->size` and `fobj->num_pages` is a safety net too, you can't map more than what they describe, no accidental crossing of borders.

If you use a fobj you need to wrap it in a mobj with `mobj_with_fobj_alloc()` since as you point out `vm_map()` only take a mobj. We are using mobj in the vm_region structure so there's currently no way around the mobj requirement.

Reply:
So instead of having one mobj/fobj for all the Normal World shares, I would create a new fobj mobj pair for each new Normal World share?

Reply:
Yes

Reply:
I changed it to use the mobj_ffa interface instead